binrw/src/container.rs

```diff
@@ -0,0 +1,121 @@
+//! container module
+#[cfg(not(feature = "std"))]
+use alloc::vec::Vec;
+
+/// Container
+pub trait Container: Sized + IntoIterator {
+    /// Count
+    type Count;
+
+    /// naive
+    ///
+    /// # Errors
+    ///
+    /// If `f` returns an error, the error will be returned
+    fn new_naive<Fun, Error>(count: Self::Count, f: Fun) -> Result<Self, Error>
+    where
+        Fun: FnMut() -> Result<Self::Item, Error>;
+
+    /// smart
+    ///
+    /// # Errors
+    ///
+    /// If `f` returns an error, the error will be returned
+    fn new_smart<Fun, Error>(count: Self::Count, f: Fun) -> Result<Self, Error>
+    where
+        Fun: FnMut(&mut [Self::Item]) -> Result<(), Error>,
+        Self::Item: Default + Clone;
```

Comment:
If features [`read_buf`](https://github.com/rust-lang/rust/issues/78485) and [`core_io_borrowed_buf`](https://github.com/rust-lang/rust/issues/117693) were stable, passing a `BorrowedCursor` to the closure instead of `&mut [Self::Item]`, and using `Read::read_buf_exact` would *almost* get us to not needing to initialize our buffers. But, with the current state of that API, we'd have no way to own the filled values (without unsafe) -- we would only get a reference to a slice of the filled values. (This doesn't mean that such an API won't eventually be added, only that it doesn't currently exist.)